shadow: avoid document mutation

at-wat opened the pull request:
Deepcopy before returning documents.
Use mutex on update.

~after #125~

at-wat commented:
@kt3k @t-asaka please take a look

kt3k commented:
When I run `example/shadow` command, I saw the following error:

```
> delete
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1348bc6]

goroutine 1 [running]:
github.com/seqsense/aws-iot-device-sdk-go/v4/shadow.(*ThingDocument).clone(0x0, 0x0)
	/Users/kt3k/sq/aws-iot-device-sdk-go/shadow/copy.go:51 +0x36
github.com/seqsense/aws-iot-device-sdk-go/v4/shadow.(*shadow).Document(0xc0001f4690, 0x0)
	/Users/kt3k/sq/aws-iot-device-sdk-go/shadow/shadow.go:383 +0x76
main.main()
	/Users/kt3k/sq/aws-iot-device-sdk-go/examples/shadow/main.go:112 +0xb2c
```
Maybe we need to check nil in `Document()` function of shadow impl?

at-wat commented:
After Delete, thing shadow document doesn't exist and AWS IoT returns error against GetDocument. I think representing it by `nil` document is reasonable.
I have fixed clone of nil document and added some tests.

kt3k commented on shadow/copy_test.go:
```diff
@@ -0,0 +1,103 @@
+	"testing"
+)
+
+func TestClone(t *testing.T) {
```
👍
